packages/typescript-estree/tools/test-utils.ts

```diff
@@ -84,3 +68,67 @@ export function isJSXFileType(fileType: string): boolean {
   }
   return fileType === 'js' || fileType === 'jsx' || fileType === 'tsx';
 }
+
+/**
+ * Removes the given keys from the given AST object recursively
+ * @param root A JavaScript object to remove keys from
+ * @param keysToOmit Names and predicate functions use to determine what keys to omit from the final object
+ * @param nodes advance ast modifications
+ * @returns formatted object
+ */
+export function omitDeep<T = Record<string, unknown>>(
+  root: T,
+  keysToOmit: { key: string; predicate: (value: unknown) => boolean }[] = [],
+  nodes: Record<string, (node: T, parent: T | null) => void> = {},
+): T {
+  function isObjectLike(value: unknown | null): value is T {
+    return (
+      typeof value === 'object' && !(value instanceof RegExp) && value !== null
+    );
+  }
+
+  function shouldOmit(keyName: string, val: unknown): boolean {
+    if (keysToOmit?.length) {
+      return keysToOmit.some(
+        keyConfig => keyConfig.key === keyName && keyConfig.predicate(val),
+      );
+    }
+    return false;
+  }
+
+  function visit(oNode: T, parent: T | null): T {
+    if (!Array.isArray(oNode) && !isObjectLike(oNode)) {
+      return oNode;
+    }
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const node: any = Array.isArray(oNode) ? [...oNode] : { ...oNode };
```

Comment:
why are we iterating over arrays here?
Should `visit` only be called on objects?

If we do `visit` objects, wouldn't it be better to do something like this, instead of introducing an `any`

```ts
if (Array.isArray(oNode)) {
  return oNode.map(visit);
}
```

Reply:
i didn't wanted to change to much of this logic, this if is actually not needed, i just pushed changes to this to refactor this a little.